lib/src/source/hosted.dart

```diff
@@ -868,23 +991,63 @@ class HostedSource extends CachedSource {
     // query-string as is the case with signed S3 URLs. And we wish to allow for
     // such URLs to be used.
     final versions = await _scheduler.schedule(_RefAndCache(id.toRef(), cache));
-    final versionInfo = versions![id];
+    final versionInfo =
+        versions.firstWhereOrNull((i) => i.version == id.version);
     final packageName = id.name;
     final version = id.version;
+    late Uint8List contentHash;
     if (versionInfo == null) {
       throw PackageNotFoundException(
           'Package $packageName has no version $version');
     }
 
     final archiveUrl = versionInfo.archiveUrl;
     log.io('Get package from $archiveUrl.');
-    log.message('Downloading ${log.bold(id.name)} ${id.version}...');
+    log.fine('Downloading ${log.bold(id.name)} ${id.version}...');
 
     // Download and extract the archive to a temp directory.
-    await withTempDir((tempDirForArchive) async {
+    return await withTempDir((tempDirForArchive) async {
       var fileName = '$packageName-$version.tar.gz';
       var archivePath = p.join(tempDirForArchive, fileName);
 
+      Stream<List<int>> validateSha256(
+        Stream<List<int>> stream,
+        Digest? expectedHash,
+      ) async* {
+        final output = _SingleValueSink<Digest>();
+        final input = sha256.startChunkedConversion(output);
+        await for (final v in stream) {
+          input.add(v);
+          yield v;
+        }
+        input.close();
+        final actualHash = output.value;
+        if (expectedHash != null && output.value != expectedHash) {
+          log.fine(
+              'Expected content-hash for ${id.name}-${id.version} $expectedHash actual: ${output.value}.');
+          throw FormatException('''
+Downloaded archive for ${id.name}-${id.version} had wrong content-hash.
+
+This indicates a problem on the package repository: `${description.url}`.
+
+See $contentHashesDocumentationUrl.
+''');
```

Comment:
This is the only thing I'm unsure of...

Options:
 * (A) Retry the request, this makes sense if the response didn't have a crc32 checksum.
 * (B) Don't retry the request, this makes sense if the response had a crc32 checksum.

But I don't like making logic that depends on whether we have crc32 checksum or not.

(C) If response carries a header `*-pub-content-sha256: <hex sha256>` then we abort fetching if the header doesn't match the `archive_sha256` from the version-listing API. If the downloaded content doesn't match the value of `*-pub-content-sha256` header, then we retry downloading.
If there is no `*-pub-content-sha256` and `archive_sha256` mismatches downloaded content, then we retry.

This is a bit of a hack, because the header will have to be anything that ends with `-pub-content-sha256`. Because on GCS we can attach custom headers, they are just prefixed `x-goog-meta-`, so our custom headers would be `x-goog-meta-pub-content-sha256` and if hosted on S3 a package repository would have `x-amx-meta-pub-content-sha256` or something like that.

See:
 * https://cloud.google.com/storage/docs/xml-api/reference-headers#xgoogmeta
 * https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingMetadata.html

--------------

But option (C) might also be overkill. Certainly it's something we should do in a follow up PR.

------------------

Back on topic, it's tempting to retry, because:
 * Most likely it's a network issues -- rarely should there be disagreement between version-listing API and the archive stored. So if there is disagreement, it's very possibly a network issue.
 * In cases where there is disagreement, its possible that it would resolve from retrying -- it's likely a temporary inconsistency in cloud APIs -- like updating a GCS object cached in a CDN might take some time to propagate.
 * If the disagreement persist, then it's a bug in the package repository, if the user have to wait for retries to take place before being told that the content-hash of the downloaded archive is wrong, then maybe that's not the end of the world. It's not like the user is waiting for a positive outcome -- it's not a common scenario.

So when in doubt, maybe it's better to retry :D